tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -382,6 +391,11 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
             frChain->Add((thisFriendFileNames[j] + "?#" + thisFriendChainSubNames[j]).c_str(), thisFriendEntries[j]);
          }
       }
+
+      const auto &indexInfo = finfo.fTreeIndexInfos[i];
+      if (indexInfo.fHasIndexInfo)
+         frChain->BuildIndex(indexInfo.fMajorName.c_str(), indexInfo.fMinorName.c_str());
```

Comment:
If `frChain` is indeed a chain, this implies that every single tree (and file) of the chain will be traverse 'right here and now', is that the intent?

Reply:
uhm I guess that instead of storing the majorname and minorname and call `BuildIndex` again in every task I could store the TTreeIndex itself and attach a copy of it to every chain in every task?

Reply:
I couldn't manage to make it work, partly because `TChainIndex::SetTree` does not actually set `fTree`, partly because I'm not sure what happens if I copy a TTreeIndex out of a TTree and then assign it to a TChain.

Reply:
> if I copy a TTreeIndex out of a TTree and then assign it to a TChain.

That does not seem to make sense, wouldn't the entry be off or not enough?

>  again in every task I could store the TTreeIndex 

For a tree index, you could, for example `TTree uses`:
```
               TVirtualIndex *index = (TVirtualIndex*) oldtree->GetTree()->GetTreeIndex()->Clone();
               index->SetTree(newtree);
               newtree->SetTreeIndex(index);
```
For a `TChainIndex` that is more dubious (beside that `TChainIndex::SetTree` would have be actually implemented) because apriori the index is very large (because it ought to an an entry for every single entry of all the trees in the chian, isn't it?) and duplicating it sounds ... less than optimal.

Reply:
`TChainIndex`, looks like it contains 'just' the range of the trees and the index for those trees that do not already have an index (so it could still be pretty large).  Ideally those would be created only once (in a distributed fashion if possible) and shared.  (It will requires updates to `TChainIndex` to make it so).

Reply:
mmh that sounds way too time consuming...

Reply:
`TChainIndex::SetTree` was changed to actually set the tree and a new version of the fix that simply clones the original TTreeIndex/TChainIndex instead of re-building every time has been pushed. Things seem to work.

I think we are missing a test for the corner case in which we have a TTree in input rather than a TChain, I'll add it soon.

Reply:
...tests added